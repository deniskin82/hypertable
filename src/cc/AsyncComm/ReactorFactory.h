/*
 * Copyright (C) 2007-2013 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/** @file
 * Declarations for ReactorFactory.
 * This file contains type declarations for ReactorFactory, a static
 * class used to create and manage I/O reactor threads.
 */


#ifndef HYPERTABLE_REACTORFACTORY_H
#define HYPERTABLE_REACTORFACTORY_H

#include <boost/random.hpp>
#include <boost/random/uniform_01.hpp>
#include <boost/thread/mutex.hpp>
#include <boost/thread/thread.hpp>

#include <cassert>
#include <set>
#include <vector>

#include "Common/atomic.h"
#include "Reactor.h"


namespace Hypertable {

  /** @addtogroup AsyncComm
   *  @{
   */

  /** Static class used to setup and manage I/O reactors.  Since the I/O reactor
   * threads are a process-wide resource, the methods of this class are static.
   */
  class ReactorFactory {

  public:

    /** Initializes I/O reactors.  This method creates and initializes
     * <code>reactor_count</code> reactors, plus an additional dedicated timer
     * reactor.  It also initializes the #use_poll member based on the
     * <code>Comm.UsePoll</code> property and sets the #ms_epollet
     * ("edge triggered") flag to <i>false</i> if running on Linux version older
     * than 2.6.17.  It also allocates a HandlerMap and initializes
     * ReactorRunner::handler_map to point to it.
     * @param reactor_count number of reactor threads to create
     */
    static void initialize(uint16_t reactor_count);

    /** This method shuts down the reactors
     */
    static void destroy();

    /** This method returns the 'next' reactor.  It returns pointers to
     * reactors in round-robin fashion and is used by the Comm subsystem to
     * evenly distribute descriptors across all of the reactors.  The
     * atomic integer variable #ms_next_reactor is used to keep track
     * of the next reactor in the list.
     * @param reactor Smart pointer reference to returned Reactor
     */
    static void get_reactor(ReactorPtr &reactor) {
      assert(ms_reactors.size() > 0);
      reactor = ms_reactors[atomic_inc_return(&ms_next_reactor)
                            % (ms_reactors.size() - 1)];
    }

    /** This method returns the timer reactor.
     * @param reactor Smart pointer reference to returned Reactor
     */
    static void get_timer_reactor(ReactorPtr &reactor) {
      assert(ms_reactors.size() > 0);
      reactor = ms_reactors.back();
    }

    /// Vector of reactors (last position is timer reactor)
    static std::vector<ReactorPtr> ms_reactors;

    /// Boost thread_group for managing reactor threads
    static boost::thread_group ms_threads;

    static boost::mt19937 rng; //!< Pseudo random number generator
    static bool ms_epollet;    //!< Use "edge triggered" epoll
    static bool use_poll;      //!< Use POSIX poll() as polling mechanism

    /// Set to <i>true</i> if this process is acting as "Proxy Master"
    static bool proxy_master;

  private:

    /// Mutex to serialize calls to #initialize
    static Mutex        ms_mutex;

    /// Atomic integer used for round-robin assignment of reactors
    static atomic_t ms_next_reactor;

  };
  /** @}*/
}

#endif // HYPERTABLE_REACTORFACTORY_H

