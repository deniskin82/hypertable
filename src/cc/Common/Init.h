/** -*- C++ -*-
 * Copyright (C) 2007-2012 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Hypertable. If not, see <http://www.gnu.org/licenses/>
 */

#ifndef HYPERTABLE_INIT_H
#define HYPERTABLE_INIT_H

#include "Common/Config.h"
#include "Common/System.h"

namespace Hypertable { namespace Config {

  /**
   * Init with policy (with init_options (called before parse_args) and
   * init (called after parse_args) methods
   *
   * See parse_args for params
   */
  template <class PolicyT>
  inline void init_with_policy(int argc, char *argv[], const Desc *desc = 0) {
    try {
      System::initialize();

      ScopedRecLock lock(rec_mutex);
      properties = new Properties();

      if (desc)
        cmdline_desc(*desc);

      file_loaded = false;
      PolicyT::init_options();
      parse_args(argc, argv);
      PolicyT::init();
      sync_aliases(); // init can generate more aliases
      //properties->notify();

      if (get_bool("verbose"))
        properties->print(std::cout);
    }
    catch (Exception &e) {
      PolicyT::on_init_error(e);
    }
  }

  /**
   * Convenience function (more of a demo) to init with a list of polices
   * @see init_with
   */
  template <class PolicyListT>
  inline void init_with_policies(int argc, char *argv[], const Desc *desc = 0) {
    typedef typename Join<PolicyListT>::type Combined;
    init_with_policy<Combined>(argc, argv, desc);
  }

  /**
   * Init with default policy
   */
  inline void init(int argc, char *argv[], const Desc *desc = NULL) {
    init_with_policy<DefaultPolicy>(argc, argv, desc);
  }

}} // namespace Hypertable::Config

#endif /* HYPERTABLE_INIT_H */
