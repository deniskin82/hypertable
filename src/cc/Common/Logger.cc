/**
 * Copyright (C) 2007-2012 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"

#include <iostream>
#include <log4cpp/Appender.hh>
#include <log4cpp/BasicLayout.hh>
#include <log4cpp/FileAppender.hh>
#include <log4cpp/Layout.hh>
#include <log4cpp/NDC.hh>
#include <log4cpp/Priority.hh>

#include "Logger.h"

using namespace Hypertable;
namespace Logging = log4cpp;

namespace {

  /**
   * NoTimeLayout
   **/
  class  NoTimeLayout : public Logging::Layout {
  public:
    NoTimeLayout() { }
    virtual ~NoTimeLayout() { }
    virtual String format(const Logging::LoggingEvent& event) {
      return Hypertable::format("%s %s %s: %s\n",
          Logging::Priority::getPriorityName(event.priority).c_str(),
          event.categoryName.c_str(), event.ndc.c_str(),
          event.message.c_str());
    }
  };

  /**
   * MicrosecondLayout
   **/
  class  MicrosecondLayout : public Logging::Layout {
  public:
    MicrosecondLayout() { }
    virtual ~MicrosecondLayout() { }
    virtual String format(const Logging::LoggingEvent& event) {
      return Hypertable::format("%d %d %s %s %s: %s\n",
          event.timeStamp.getSeconds(), event.timeStamp.getMicroSeconds(),
          Logging::Priority::getPriorityName(event.priority).c_str(),
          event.categoryName.c_str(), event.ndc.c_str(),
          event.message.c_str());
    }
  };

  /**
   * FlushableAppender appender with flush method
   */
  class FlushableAppender : public Logging::LayoutAppender {
  public:
    FlushableAppender(const String &name) : Logging::LayoutAppender(name) {}
    virtual ~FlushableAppender() {}
    virtual void flush() = 0;

    bool
    set_flush_per_log(bool choice) {
      bool old = m_flush_per_log;
      m_flush_per_log = choice;
      return old;
    }

  protected:
    bool m_flush_per_log;
  };

  /**
   * FlushableOstreamAppender appends LoggingEvents to ostreams.
   */
  class FlushableOstreamAppender : public FlushableAppender {
  public:
    FlushableOstreamAppender(const String& name, std::ostream &stream,
                             bool flush_per_log) :
                             FlushableAppender(name), m_stream(stream) {
      set_flush_per_log(flush_per_log);
    }
    virtual ~FlushableOstreamAppender() { close(); }

    virtual bool reopen() { return true; }
    virtual void close() { flush(); }
    virtual void flush() { m_stream.flush(); }

  protected:
    virtual void
    _append(const Logging::LoggingEvent& event) {
      m_stream << _getLayout().format(event);

      if (m_flush_per_log)
        m_stream.flush();
    }

  private:
    std::ostream &m_stream;
  };

  FlushableAppender *appender = 0;

} // local namespace

Logging::Category *Logger::logger = 0;
bool Logger::show_line_numbers = true;

void
Logger::initialize(const String &name, int priority, bool flush_per_log,
                   std::ostream &out) {
  appender = new FlushableOstreamAppender("default", out, flush_per_log);
  Logging::Layout* layout = new Logging::BasicLayout();
  //Logging::Layout* layout = new MicrosecondLayout();
  appender->setLayout(layout);
  logger = &(Logging::Category::getInstance(name));
  logger->addAppender(appender);
  logger->setAdditivity(false);
  logger->setPriority(priority);
}

void
Logger::set_level(int priority) {
  logger->setPriority(priority);
}

void
Logger::suppress_line_numbers() {
  Logger::show_line_numbers = false;
}

void
Logger::set_test_mode(const String &name, int fd) {
  Logging::FileAppender *appender;
  Logger::show_line_numbers = false;
  logger->removeAllAppenders();
  appender = new Logging::FileAppender(name, fd);
  appender->setLayout(new NoTimeLayout());
  logger->setAppender(appender);
}

bool
Logger::set_flush_per_log(bool choice) {
  return appender->set_flush_per_log(choice);
}

void
Logger::flush() {
  appender->flush();
}
