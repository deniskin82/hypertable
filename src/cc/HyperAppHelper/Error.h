/**
 * Copyright (C) 2007-2012 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERAPPHELPER_ERROR_H
#define HYPERAPPHELPER_ERROR_H

#include "Common/Compat.h"
#include "Common/Error.h"

namespace Hypertable { namespace HyperAppHelper {

  /**
   * Retrieves a descriptive error string of an error code
   *
   * @param error_code The table pointer
   *
   * @returns a descriptive error string, or "ERROR NOT REGISTERED" if 
   *     if the error_code is unknown
   */
  const char *error_get_text(int error_code) {
    return Hypertable::Error::get_text(error_code);
  }

}} // namespace HyperAppHelper, HyperTable

#endif // HYPERAPPHELPER_ERROR_H

