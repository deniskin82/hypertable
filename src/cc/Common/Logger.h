/**
 * Copyright (C) 2007-2012 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */
#ifndef HYPERTABLE_LOGGER_H
#define HYPERTABLE_LOGGER_H

#include "Error.h"
#include "String.h"
#include <iostream>
#include "FixedStream.h"
#include <log4cpp/Category.hh>

namespace Hypertable { namespace Logger {
  using log4cpp::Priority;

  void initialize(const String &name, int level = Priority::DEBUG,
                  bool flush_per_log = true, std::ostream &out = std::cout);
  void set_level(int level);
  void set_test_mode(const String &name, int fd=1);
  void suppress_line_numbers();
  void flush();
  bool set_flush_per_log(bool);

  extern log4cpp::Category *logger;
  extern bool show_line_numbers;

}} // namespace Hypertable::Logger


#define HT_LOG_BUFSZ 1024

// This should generate a core dump
#ifdef HT_USE_ABORT
#define HT_ABORT abort()
#else
#define HT_ABORT *((int *)0) = 1
#endif

// printf interface macro helper
#define HT_LOG(_enabled_, _cat_, msg) do { \
  if (Logger::logger->_enabled_()) { \
    if (Logger::show_line_numbers) \
      Logger::logger->log(log4cpp::Priority::_cat_, Hypertable::format( \
          "(%s:%d) %s", __FILE__, __LINE__, msg)); \
    else \
      Logger::logger->log(log4cpp::Priority::_cat_, msg); \
  } \
} while (0)

#define HT_LOGF(_enabled_, _cat_, fmt, ...) do { \
  if (Logger::logger->_enabled_()) { \
    if (Logger::show_line_numbers) \
      Logger::logger->log(log4cpp::Priority::_cat_, Hypertable::format( \
          "(%s:%d) " fmt, __FILE__, __LINE__, __VA_ARGS__)); \
    else \
      Logger::logger->log(log4cpp::Priority::_cat_, Hypertable::format( \
          fmt, __VA_ARGS__));  \
  } \
} while (0)

// stream interface macro helpers
#define HT_LOG_BUF_SIZE 4096

#define HT_OUT(_enabled_, _l_) do { if (Logger::logger->_enabled_()) { \
  char logbuf[HT_LOG_BUF_SIZE]; \
  log4cpp::Priority::PriorityLevel _level_ = log4cpp::Priority::_l_; \
  FixedOstream _out_(logbuf, sizeof(logbuf)); \
  if (Logger::show_line_numbers) \
    _out_ <<"("<< __FILE__ <<':'<< __LINE__ <<") "; \
  _out_

#define HT_OUT2(_enabled_, _l_) do { if (Logger::logger->_enabled_()) { \
  char logbuf[HT_LOG_BUF_SIZE]; \
  log4cpp::Priority::PriorityLevel _level_ = log4cpp::Priority::_l_; \
  FixedOstream _out_(logbuf, sizeof(logbuf)); \
  _out_ << __func__; \
  if (Logger::show_line_numbers) \
    _out_ << " ("<< __FILE__ <<':'<< __LINE__ <<")"; \
  _out_ <<": "

#define HT_END ""; Logger::logger->log(_level_, _out_.str()); \
  if (_level_ == log4cpp::Priority::FATAL) HT_ABORT; \
} /* if enabled */ } while (0)

#define HT_OUT_DISABLED do { if (0) {

// helpers for printing a char pointer field
#define HT_DUMP_CSTR(_os_, _label_, _str_) do { \
  if (!_str_) _os_ <<" " #_label_ "=[NULL]"; \
  else _os_ <<" " #_label_ "='"<< (_str_) << "'"; \
} while (0)

#define HT_DUMP_CSTR_FIELD(_os_, _obj_, _field_) \
  HT_DUMP_CSTR(_os_, _field_, _obj_._field_)


// Logging macros interface starts here
#ifndef HT_DISABLE_LOG_ALL

#ifndef HT_DISABLE_LOG_DEBUG

#define HT_LOG_ENTER do { \
  if (Logger::logger->isDebugEnabled()) {\
    if (Logger::show_line_numbers) \
      Logger::logger->debug("(%s:%d) %s() ENTER", __FILE__, __LINE__, HT_FUNC);\
    else \
      Logger::logger->debug("%s() ENTER", HT_FUNC); \
  } \
} while(0)

#define HT_LOG_EXIT do { \
  if (Logger::logger->isDebugEnabled()) { \
    if (Logger::show_line_numbers) \
      Logger::logger->debug("(%s:%d) %s() EXIT", __FILE__, __LINE__, HT_FUNC); \
    else \
      Logger::logger->debug("%s() EXIT", HT_FUNC); \
  } \
} while(0)

#define HT_DEBUG(msg) HT_LOG(isDebugEnabled, DEBUG, msg)
#define HT_DEBUGF(msg, ...) HT_LOGF(isDebugEnabled, DEBUG, msg, __VA_ARGS__)
#define HT_DEBUG_OUT HT_OUT2(isDebugEnabled, DEBUG)
#else
#define HT_LOG_ENTER
#define HT_LOG_EXIT
#define HT_DEBUG(msg)
#define HT_DEBUGF(msg, ...)
#define HT_DEBUG_OUT HT_OUT_DISABLED
#endif

#ifndef HT_DISABLE_LOG_INFO
#define HT_INFO(msg) HT_LOG(isInfoEnabled, INFO, msg)
#define HT_INFOF(msg, ...) HT_LOGF(isInfoEnabled, INFO, msg, __VA_ARGS__)
#define HT_INFO_OUT HT_OUT(isInfoEnabled, INFO)
#else
#define HT_INFO(msg)
#define HT_INFOF(msg, ...)
#define HT_INFO_OUT HT_OUT_DISABLED
#endif

#ifndef HT_DISABLE_LOG_NOTICE
#define HT_NOTICE(msg) HT_LOG(isNoticeEnabled, NOTICE, msg)
#define HT_NOTICEF(msg, ...) HT_LOGF(isNoticeEnabled, NOTICE, msg, __VA_ARGS__)
#define HT_NOTICE_OUT HT_OUT(isNoticeEnabled, NOTICE)
#else
#define HT_NOTICE(msg)
#define HT_NOTICEF(msg, ...)
#define HT_NOTICE_OUT HT_OUT_DISABLED
#endif

#ifndef HT_DISABLE_LOG_WARN
#define HT_WARN(msg) HT_LOG(isWarnEnabled, WARN, msg)
#define HT_WARNF(msg, ...) HT_LOGF(isWarnEnabled, WARN, msg, __VA_ARGS__)
#define HT_WARN_OUT HT_OUT2(isWarnEnabled, WARN)
#else
#define HT_WARN(msg)
#define HT_WARNF(msg, ...)
#define HT_WARN_OUT HT_OUT_DISABLED
#endif

#ifndef HT_DISABLE_LOG_ERROR
#define HT_ERROR(msg) HT_LOG(isErrorEnabled, ERROR, msg)
#define HT_ERRORF(msg, ...) HT_LOGF(isErrorEnabled, ERROR, msg, __VA_ARGS__)
#define HT_ERROR_OUT HT_OUT2(isErrorEnabled, ERROR)
#else
#define HT_ERROR(msg)
#define HT_ERRORF(msg, ...)
#define HT_ERROR_OUT HT_OUT_DISABLED
#endif

#ifndef HT_DISABLE_LOG_CRIT
#define HT_CRIT(msg) HT_LOG(isCritEnabled, CRIT, msg)
#define HT_CRITF(msg, ...) HT_LOGF(isCritEnabled, CRIT, msg, __VA_ARGS__)
#define HT_CRIT_OUT HT_OUT2(isCritEnabled, CRIT)
#else
#define HT_CRIT(msg)
#define HT_CRITF(msg, ...)
#define HT_CRIT_OUT HT_OUT_DISABLED
#endif

#ifndef HT_DISABLE_LOG_ALERT
#define HT_ALERT(msg) HT_LOG(isAlertEnabled, ALERT, msg)
#define HT_ALERTF(msg, ...) HT_LOGF(isAlertEnabled, ALERT, msg, __VA_ARGS__)
#define HT_ALERT_OUT HT_OUT2(isAlertEnabled, ALERT)
#else
#define HT_ALERT(msg)
#define HT_ALERTF(msg, ...)
#define HT_ALERT_OUT HT_OUT_DISABLED
#endif

#ifndef HT_DISABLE_LOG_EMERG
#define HT_EMERG(msg) HT_LOG(isEmergEnabled, EMERG, msg)
#define HT_EMERGF(msg, ...) HT_LOGF(isEmergEnabled, EMERG, msg, __VA_ARGS__)
#define HT_EMERG_OUT HT_OUT2(isEmergEnabled, EMERG)
#else
#define HT_EMERG(msg)
#define HT_EMERGF(msg, ...)
#define HT_EMERG_OUT HT_OUT_DISABLED
#endif

#ifndef HT_DISABLE_LOG_FATAL
#define HT_FATAL(msg) do { \
  HT_LOG(isFatalEnabled, FATAL, msg); \
  HT_ABORT; \
} while (0)
#define HT_FATALF(msg, ...) do { \
  HT_LOGF(isFatalEnabled, FATAL, msg, __VA_ARGS__); \
  HT_ABORT; \
} while (0)
#define HT_FATAL_OUT HT_OUT2(isFatalEnabled, FATAL)
#else
#define HT_FATAL(msg)
#define HT_FATALF(msg, ...)
#define HT_FATAL_OUT HT_OUT_DISABLED
#endif

#else // HT_DISABLE_LOGGING

#define HT_DEBUG(msg)
#define HT_DEBUGF(msg, ...)
#define HT_INFO(msg)
#define HT_INFOF(msg, ...)
#define HT_NOTICE(msg)
#define HT_NOTICEF(msg, ...)
#define HT_WARN(msg)
#define HT_WARNF(msg, ...)
#define HT_ERROR(msg)
#define HT_ERRORF(msg, ...)
#define HT_CRIT(msg)
#define HT_CRITF(msg, ...)
#define HT_ALERT(msg)
#define HT_ALERTF(msg, ...)
#define HT_EMERG(msg)
#define HT_EMERGF(msg, ...)
#define HT_FATAL(msg)
#define HT_FATALF(msg, ...)
#define HT_LOG_ENTER
#define HT_LOG_EXIT
#define HT_DEBUG_OUT HT_OUT_DISABLED
#define HT_INFO_OUT HT_OUT_DISABLED
#define HT_NOTICE_OUT HT_OUT_DISABLED
#define HT_WARN_OUT HT_OUT_DISABLED
#define HT_ERROR_OUT HT_OUT_DISABLED
#define HT_CRIT_OUT HT_OUT_DISABLED
#define HT_ALERT_OUT HT_OUT_DISABLED
#define HT_EMERG_OUT HT_OUT_DISABLED
#define HT_FATAL_OUT HT_OUT_DISABLED

#endif // HT_DISABLE_LOGGING

// Probably should be in its own file, but...
#define HT_EXPECT(_e_, _code_) do { if (_e_); else { \
    if (_code_ == Error::FAILED_EXPECTATION) \
      HT_FATAL("failed expectation: " #_e_); \
    HT_THROW(_code_, "failed expectation: " #_e_); } \
} while (0)

// A short cut for HT_EXPECT(expr, Error::FAILED_EXPECTATION)
// unlike assert, it cannot be turned off
#define HT_ASSERT(_e_) HT_EXPECT(_e_, Error::FAILED_EXPECTATION)

#endif // HYPERTABLE_LOGGER_H
