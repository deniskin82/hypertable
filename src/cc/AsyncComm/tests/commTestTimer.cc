/**
 * Copyright (C) 2007-2012 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include <cstdlib>

extern "C" {
#include <poll.h>
#include <signal.h>
#include <sys/types.h>
#include <unistd.h>
}

#include <boost/thread/thread.hpp>

#include "Common/Init.h"
#include "Common/Error.h"
#include "Common/FileUtils.h"
#include "Common/Logger.h"
#include "Common/TestHarness.h"
#include "Common/System.h"
#include "Common/Usage.h"

#include "AsyncComm/Comm.h"
#include "AsyncComm/ConnectionManager.h"
#include "AsyncComm/Event.h"
#include "AsyncComm/ReactorFactory.h"

#include "CommTestThreadFunction.h"

using namespace std;
using namespace Hypertable;

namespace {
  const char *usage[] = {
    "usage: commTestTimer",
    "",
    "This program tests tests timer logic of AsyncComm.",
    0
  };

  /**
   *
   */
  class TimerHandler : public DispatchHandler {
  public:
    TimerHandler(const char *msg)
      : m_msg(msg) { return; }
    virtual void handle(EventPtr &event_ptr) {
      HT_INFOF("%s (%s)", event_ptr->to_str().c_str(), m_msg);
    }
  private:
    const char *m_msg;
  };

  struct TimerRec {
    uint32_t delay;
    const char *msg;
  };

  struct TimerRec history[] = {
    { 1, "Local" },
    { 2, "media" },
    { 3, "have" },
    { 4, "reported" },
    { 5, "eyewitness" },
    { 6, "accounts" },
    { 7, "of" },
    { 8, "a" },
    { 9, "fiery" },
    { 10, "ball" },
    { 35, 0 }
  };
}


int main(int argc, char **argv) {
  Comm *comm;
  int error;
  TestHarness harness("commTestTimer");
  bool golden = false;
  TimerHandler *timer_handler;
  uint32_t wait_time = 0;

  Config::init(argc, argv);

  if (argc > 1) {
    if (!strcmp(argv[1], "--golden"))
      golden = true;
    else
      Usage::dump_and_exit(usage);
  }

  srand(8876);

  ReactorFactory::initialize(1);

  comm = Comm::instance();

  for (int i=0; history[i].msg; i++) {
    timer_handler = new TimerHandler(history[i].msg);
    if ((error = comm->set_timer(history[i].delay*1000, timer_handler))
        != Error::OK) {
      HT_ERRORF("Problem setting timer - %s", Error::get_text(error));
      exit(1);
    }
    if (history[i].delay > wait_time)
      wait_time = history[i].delay;
  }

  poll(0, 0, (wait_time+1)*1000);

  if (!golden)
    harness.validate_and_exit("commTestTimer.golden");

  harness.regenerate_golden_file("commTestTimer.golden");

  _exit(0);
}
