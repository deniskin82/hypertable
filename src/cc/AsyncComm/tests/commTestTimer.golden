INFO commTestTimer : Event: type=TIMER (Local)
INFO commTestTimer : Event: type=TIMER (media)
INFO commTestTimer : Event: type=TIMER (have)
INFO commTestTimer : Event: type=TIMER (reported)
INFO commTestTimer : Event: type=TIMER (eyewitness)
INFO commTestTimer : Event: type=TIMER (accounts)
INFO commTestTimer : Event: type=TIMER (of)
INFO commTestTimer : Event: type=TIMER (a)
INFO commTestTimer : Event: type=TIMER (fiery)
INFO commTestTimer : Event: type=TIMER (ball)
